File: Programmers/targetnumber.cpp
#include <string>
#include <vector>

using namespace std;

int N;
int count;

void dfs(vector<int> numbers, int target, int sum, int k){
	if(k == N) {
		if(target == sum){
			count++;
			printf("%d\n", count);
		}
		return;
	} 
	else {
		sum += numbers[k];
		dfs(numbers, target, sum, k+1);
		sum -= numbers[k];
		sum -= numbers[k];
		dfs(numbers, target, sum, k+1);
	}
}

int solution(vector<int> numbers, int target) {
    N = numbers.size();
    dfs(numbers, target, 0, 0);
    return count;
}


vector<int> param{1, 1, 1, 1, 1};
int main() {
	solution(param, 3);
	
	return 0;
}

File: Programmers/old/BFS/targetNumber.py
# 타겟 넘버
# From : https://programmers.co.kr/learn/courses/30/lessons/43165
#

def solution(numbers, target):
    print(numbers, target)
    if not numbers and target == 0 :
        return 1
    elif not numbers:
        return 0
    else:
        return solution(numbers[1:], target-numbers[0]) + solution(numbers[1:], target+numbers[0])


print(solution([1, 1, 1, 1, 1],3))